packages/pinning-crust-backend/src/index.ts

```diff
@@ -0,0 +1,205 @@
+import { toString } from 'uint8arrays/to-string'
+import { KeyringPair } from '@polkadot/keyring/types';
+import { DispatchError } from '@polkadot/types/interfaces';
+import { ITuple } from '@polkadot/types/types';
+import { SubmittableExtrinsic } from '@polkadot/api/promise/types';
+
+// Encoder
+const textEncoder = new TextEncoder()
+
+// Errors
+export class EmptySeedError extends Error {
+  constructor(address: string) {
+    super(`[Crust]: No seed provided at ${address}`)
+  }
+}
+
+export class TxError extends Error {
+  constructor(type: number, msg: string) {
+    super(`[Crust]: Send transaction(${type}) failed with ${msg}`)
+  }
+}
+
+// Object
+interface ListRes {
+  data: {
+    substrate_extrinsic: Array<{ args: string }>
+  }
+}
+
+// Pinning class
+export class CrustPinningBackend implements PinningBackend {
+  static designator = 'crust'
+
+  readonly id: string
+  readonly endpoint: string
+  readonly seed: string
+  api: ApiPromise
+
+  constructor(readonly connectionString: string) {
+    // Check input
+    const url = new URL(connectionString)
+    const hostname = url.hostname
+    const seed = url.searchParams.get('seed')
+    if (!seed) {
+      throw new EmptySeedError(this.endpoint)
+    }
+    this.seed = seed
+
+    const protocol = url.protocol
+      .replace('crust:', 'ws')
+      .replace('crust+ws:', 'ws')
+      .replace('crust+wss:', 'wss')
+
+    if (url.port) {
+      this.endpoint = `${protocol}://${hostname}:${url.port}`
+    } else {
+      this.endpoint = `${protocol}://${hostname}`
+    }
+
+    // ID
+    const bytes = textEncoder.encode(this.connectionString)
+    const digest = toString(sha256.hash(bytes), 'base64urlpad')
+    this.id = `${CrustPinningBackend.designator}@${digest}`
+  }
+
+  open(): void {
+    this.api = new ApiPromise({
+      provider: new WsProvider(this.endpoint),
+      typesBundle: typesBundleForPolkadot,
+    });
+  }
+
+  async close(): Promise<void> {
+    if (this.api) {
+      await this.api.disconnect()
+      this.api = undefined
+    }
+  }
+
+  async pin(cid: CID): Promise<void> {
+    // Check connnetion
+    if (!this.api) {
+      return
+    }
+    await this.api.isReadyOrError
+
+    // Get a onchain keypair
+    const kr = new Keyring({ type: 'sr25519', ss58Format: 66 })
+    const krp = kr.addFromUri(this.seed)
+
+    // Storage order
+    // Support for the storage of up to 30G
+    const size = 30 * 1024 * 1024 * 1024
+    const tx = this.api.tx.market.placeStorageOrder(cid.toString(), size, 0, '')
+    await this.sendTx(tx, krp)
+
+    // Permanent storage
+    // Learn what's prepard for: https://wiki.crust.network/docs/en/DSM#3-file-order-assurance-settlement-and-discount
+    // In pCRU, 1 pCRU = 10^-12 CRU
+    const prepard = 1000000000 // in pCRU, 1 pCRU = 10^-12 CRU
+    const tx2 = this.api.tx.market.addPrepaid(cid.toString(), prepard)
+    await this.sendTx(tx2, krp)
+  }
+
+  async unpin(_cid: CID): Promise<void> {
```

Comment:
Is CRUST a permanent storage? Is `unpin` function unsupported on the network?

Reply:
Yes, Crust is permanent storage

Reply:
Good to know, thanks.